js/packages/mobile-wallet-adapter-walletlib/src/useMobileWalletAdapterSession.ts

```diff
@@ -0,0 +1,76 @@
+import { TransactionVersion } from '@solana/web3.js';
+import { useEffect } from 'react';
+import { NativeEventEmitter, NativeModules, Platform } from 'react-native';
+
+import { MWASessionEvent, MWASessionEventType } from './mwaSessionEvents.js';
+import { MWARequest, MWARequestType } from './resolve.js';
+
+const LINKING_ERROR =
+    `The package 'solana-mobile-wallet-adapter-walletlib' doesn't seem to be linked. Make sure: \n\n` +
+    '- You rebuilt the app after installing the package\n' +
+    '- If you are using Lerna workspaces\n' +
+    '  - You have added `@solana-mobile/mobile-wallet-adapter-walletlib` as an explicit dependency, and\n' +
+    '  - You have added `@solana-mobile/mobile-wallet-adapter-walletlib` to the `nohoist` section of your package.json\n' +
+    '- You are not using Expo managed workflow\n';
+
+const SolanaMobileWalletAdapterWalletLib =
+    Platform.OS === 'android' && NativeModules.SolanaMobileWalletAdapterWalletLib
+        ? NativeModules.SolanaMobileWalletAdapterWalletLib
+        : new Proxy(
+              {},
+              {
+                  get() {
+                      throw new Error(
+                          Platform.OS !== 'android'
+                              ? 'The package `solana-mobile-wallet-adapter-walletlib` is only compatible with React Native Android'
+                              : LINKING_ERROR,
+                      );
+                  },
+              },
+          );
+
+const MOBILE_WALLET_ADAPTER_EVENT_BRIDGE_NAME = 'MobileWalletAdapterServiceRequestBridge';
+
+export interface MobileWalletAdapterConfig {
+    supportsSignAndSendTransactions: boolean;
+    maxTransactionsPerSigningRequest: number;
+    maxMessagesPerSigningRequest: number;
+    supportedTransactionVersions: Array<TransactionVersion>;
+    noConnectionWarningTimeoutMs: number;
+}
+
+export function useMobileWalletAdapterSession(
+    walletName: string,
+    config: MobileWalletAdapterConfig,
+    handleRequest: (request: MWARequest) => void,
+    handleSessionEvent: (sessionEvent: MWASessionEvent) => void,
+) {
+    // Start native event listeners
+    useEffect(() => {
+        const mwaEventEmitter = new NativeEventEmitter();
+        mwaEventEmitter.addListener(MOBILE_WALLET_ADAPTER_EVENT_BRIDGE_NAME, (nativeEvent) => {
```

Comment:
I thought we were planning to break this up into 2 separate emitters/listeners, one for requests, one for session events.

Reply:
If we had a separate event emitter for the session callbacks, could we make it so the consumer can subscribe based on the session id? so one listener per session

Reply:
The plan was to just make a type distinction between events and requests. I was just planning on distinguishing through `__type` like @steveluscher  mentioned above. Having two separate native event listeners works too, but not sure if there are any performance implications, i'm not seeing a huge advantage to having 2.

> could we make it so the consumer can subscribe based on the session id

The dev can still do this within the `handleSessionEvent` callback they pass in. How they choose to implement that is up to them. i.e:
```
// super dumb implementation. In practice they'd use some object/map to keep track of sessions.
const handleSessionEvent = (sessionEvent) => {
    if (sessionEvent.sessionId === '123') {
          // handle events from '123'
    } else if (sessionEvent.sessionId === '124') {
         // handle events from '124'
    }
}